solrsearch-plugins/src/main/java/co.cask.hydrator.plugin/batch/SolrSearchSink.java

```diff
@@ -0,0 +1,160 @@
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.cdap.format.StructuredRecordStringConverter;
+import co.cask.hydrator.plugin.common.SolrOutputFormat;
+import co.cask.hydrator.plugin.common.SolrRecordWriter;
+import co.cask.hydrator.plugin.common.SolrSearchSinkConfig;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.apache.hadoop.io.Text;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Batch SolrSearch Sink Plugin - Writes data to a SingleNode Solr or to SolrCloud.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name("SolrSearch")
+@Description("This plugin allows users to build the pipelines to write data to Solr. The input fields coming from " +
+  "the previous stage of the pipeline are mapped to Solr fields. User can also specify the mode of the Solr to " +
+  "connect to. For example, SingleNode Solr or SolrCloud.")
+public class SolrSearchSink extends BatchSink<StructuredRecord, Text, Text> {
+  private static final Gson GSON = new Gson();
+  private static final Type SCHEMA_TYPE = new TypeToken<Schema>() { }.getType();
+  private final BatchSolrSearchConfig batchConfig;
+  private StageMetrics metrics;
+
+  public SolrSearchSink(BatchSolrSearchConfig batchConfig) {
+    this.batchConfig = batchConfig;
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws Exception {
+    context.addOutput(Output.of(batchConfig.referenceName, new SolrSearchSink.SolrOutputFormatProvider(batchConfig)));
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    batchConfig.validateSolrConnectionString();
+    if (inputSchema != null) {
+      batchConfig.validateKeyField(inputSchema);
+      batchConfig.validateInputFieldsDataType(inputSchema);
+    }
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    metrics = context.getMetrics();
+    /* Calling testSolrConnection() before each mapper, to ensure that the connection is alive and available for
+    indexing.*/
+    batchConfig.testSolrConnection();
+  }
+
+  @Override
+  public void transform(StructuredRecord structuredRecord, Emitter<KeyValue<Text, Text>> emitter) throws Exception {
+    Schema recordSchema = structuredRecord.getSchema();
+    batchConfig.validateKeyField(recordSchema);
+    batchConfig.validateInputFieldsDataType(recordSchema);
+
+    if (structuredRecord.get(batchConfig.getKeyField()) == null) {
+      metrics.count("invalid", 1);
+      return;
+    }
+    emitter.emit(new KeyValue<Text, Text>(new Text(GSON.toJson(recordSchema, SCHEMA_TYPE)),
+                                          new Text(StructuredRecordStringConverter.toJsonString(structuredRecord))));
```

Comment:
Kept type to org.apache.hadoop.io.Text as the SolrOutputFormat and SolrRecordWriter classes are specifically related to Solr indexing with mapreduce functionality. And since SolrRecordWriter extends RecordWriter<Text, Text>, hence kept emitter.emit(new KeyValue<Text, Text> type to Text also. 

However, tested the plugin by changing the type to java.lang.String (in both SolrRecordWriter and SolrSearchSink classes), it worked.
Please let me know, should we change it to String type or the Text is OK.